### PR TITLE
Feature/messaging

### PR DESCRIPTION
**Pull request by baudoin-csp**
This PR introduces the initial implementation of the messaging feature between a landlord and a renter. Below is a comparison between the Figma design and the current version. Note that image handling is pending due to the camera feature, and the header is not yet included, as it was incomplete at the time of development.

<div style="display: flex; gap: 20px;">
Screenshot 2024-10-17 at 20 58 54
messaging v2
</div>

Here is a breakdown of the different steps that I went through:

### Create a valid and scalable database structure to store messages of each user
Here is a summary of the linked issue: https://github.com/Swent-Fall-2024-team-10/leazy/issues/42

The database uses a separate collection for messages, storing the UUIDs of both the renter and landlord, as well as metadata. This avoids cluttering user documents and makes filtering messages straightforward, based on the current user. A single landlord is assigned per residence, with multiple renters. Storing messages in a separate collection ensures scalability as the conversation history grows, preventing potential query slowdowns. Additionally, we want that metadata includes timestamps for message retrieval (latest first) to optimize query times.
Here is a picture of the Firestore collection so far: 

Screenshot 2024-10-17 at 20 54 13

### Display the screen in one of the tab of the menu
To easily access the first version of the messaging app, we decided to create a temporary tab in the drawer. This allows progress without depending on the "Opening an Issue" screen, which is still under development.

Screenshot 2024-10-17 at 21 08 52

### Messaging Framework: GiftedChat
The react-native-gifted-chat library was selected due to its:
- Design similarity to the Figma mockups
- Active maintenance (recent commit within the last 2 weeks)
- Full customizability (from the general appearance (background color, styling, etc...) to advanced features such as handling audio, rendering video, display avatar, "is writing" feature, etc...

What I did:
- Defined message structure and associated state.
- Ensured safe area rendering on newer iPhone models (iPhone 12 to iPhone 15).
- Customized message styling via the Bubble prop.
- Implemented message collection creation and user-based conditional filtering.

### Next steps

- We need to implement unit test and end-to-end test with multiple users and one landlord. 
- Add the possibility to send photos, videos and audio.
- Integrate the messaging feature inside the issue screen

### CODE REVIEW

File: app/index.tsx [hunk 1]
```diff
@@ -39,6 +57,7 @@ export default function App() {
         <Drawer.Screen name="Shared elements" component={SharedElementsScreen} />
         <Drawer.Screen name="Subrent" component={SubrentScreen} />
         <Drawer.Screen name="Settings" component={SettingsScreen} />
+        <Drawer.Screen name="(Temp) Messaging" component={MessagingScreen} />
       </Drawer.Navigator>
     </NavigationContainer>
```
Comment (Rcc999) [inline app/index.tsx:63]: Good the code is straightforward and follow the navigation implemented during the s1.
You will have to find a way to really make the chat appear temporarily in the navigation bar.

File: app/screens/MessagingScreen.tsx
Comment (Rcc999) [inline app/screens/MessagingScreen.tsx:1]: ### **Pull Request Review Summary for Chat Component**

#### **Strengths**
1. **Firebase Integration**: Authentication and real-time message fetching are well implemented, ensuring smooth functionality.
2. **State Management**: Managing chat messages and sign-in status via `useState` and `useLayoutEffect` is efficient.
3. **Custom Chat Bubbles**: Customizing message bubbles for sender and receiver adds to user experience.
4. **Library Choice**: The use of `react-native-gifted-chat` is a great choice, as it provides a flexible, out-of-the-box solution for chat UI, reducing the need for custom components and speeding up development.

#### **Improvements**
1. **Dynamic Authentication**: The mock user sign-in should be replaced with real authentication logic in future.

**Review by Rcc999 — APPROVED** [review]
### **General PR Review**

This PR introduces a well-implemented chat feature using `react-native-gifted-chat` and Firebase for authentication and real-time messaging. The choice of `react-native-gifted-chat` simplifies the UI while offering flexibility for future customization.

There are a few areas to improve:
- Replace the mock user sign-in with a dynamic authentication solution.
- Find a way tp really make the chat temporary

Additionally, consider adding more UI tests to ensure the chat functionality works as expected, especially in edge message sending failures.

Overall, the component is well-structured, but consider adding comments for clarity.
